staging/src/k8s.io/apiserver/pkg/storage/cacher/time_budget.go

```diff
@@ -48,43 +48,45 @@ type timeBudgetImpl struct {
 	sync.Mutex
 	budget time.Duration
 
-	refresh   time.Duration
-	maxBudget time.Duration
+	refresh         time.Duration
+	maxBudget       time.Duration
+	lastRefreshTime time.Time
 }
 
-func newTimeBudget(stopCh <-chan struct{}) timeBudget {
-	result := &timeBudgetImpl{
-		budget:    time.Duration(0),
-		refresh:   refreshPerSecond,
-		maxBudget: maxBudget,
-	}
-	go result.periodicallyRefresh(stopCh)
-	return result
-}
-
-func (t *timeBudgetImpl) periodicallyRefresh(stopCh <-chan struct{}) {
-	ticker := time.NewTicker(time.Second)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-ticker.C:
-			t.Lock()
-			if t.budget = t.budget + t.refresh; t.budget > t.maxBudget {
-				t.budget = t.maxBudget
-			}
-			t.Unlock()
-		case <-stopCh:
-			return
-		}
+func newTimeBudget() timeBudget {
+	return &timeBudgetImpl{
+		budget:          time.Duration(0),
+		refresh:         refreshPerSecond,
+		maxBudget:       maxBudget,
+		lastRefreshTime: time.Now(), // set the last refresh time to now
 	}
 }
 
 func (t *timeBudgetImpl) takeAvailable() time.Duration {
 	t.Lock()
 	defer t.Unlock()
-	result := t.budget
+
+	// freeze the time of refresh
+	now := time.Now()
```

Comment:
As discussed on the issue - please use clock.Clock.

Reply:
This is done now. 👍🏽